### DISCUSSION TITLE
Schema is generated using field alias

### DISCUSSION THREAD

**Discussion by aleksarias** [fastapi/fastapi#9150] [Questions] [answered]
### Describe the bug

Schema that is generated in FasAPI `/docs` endpoint is generated with alias rather than the model field when in the route configuration it is specified the route will reply with the model fields instead of the model fields' alias. Route for endpoint is using `response_model` to define the response schema. I tried setting `response_model_by_alias` to `False` which correctly uses the model field's instead of mode field aliases in the response however the schema is wrong in this case. 

### To Reproduce

Steps to reproduce the behavior with a minimum self-contained file.

Replace each part with your own scenario:

1. Create a file with:

```Python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

class Transaction(BaseModel):
    id: ObjectIdStr = Field(..., alias='_id')
    loan_id: ObjectIdStr = Field(..., alias='loanId')

@app.get("/", response_model=Transaction, response_model_by_alias=False)
def read_root():
    return {"Hello": "World"}
```

2. Open the browser and call the endpoint `/docs`.
3. Check schema response vs the actual response and notice schema generated is inconsistent with the response. 

### Expected behavior

Schema should use the model fields instead of using model fields' alias when `response_model_by_alias=False` is set.

### Environment

- OS: Linux / Windows / macOS
- FastAPI Version: 0.44
- Python version: 3.8

**Comment by rappongy** [comment]
@aleksarias , I cannot reproduce your error. What is the "ObjectIdStr" ?

**Comment by iishyfishyy** [comment]
@rappongy I believe what he's saying is if you have an alias in a model, the response model will transform the response using the alias regardless of the `response_model_by_alias` value. I'm not sure `ObjectIdStr` is relevant!

**Comment by JinKazuya** [comment]
Anyone found a workaround for this?

**Comment by JinKazuya** [comment]
I have found the cause of the issue in v. 0.52.0 but it can equally be found in the master branch.

The issue stems from **not** passing the `by_alias` flag to the openapi creation code when the schemas for the app's endpoints are created in this function:
https://github.com/tiangolo/fastapi/blob/0.52.0/fastapi/utils.py#L83-L85

Using OPs code, it can be reproduced with:
```
client = TestClient(app)
response = client.get('/openapi.json')
```

While this bug creates a disconnect between FastAPI and Pydantic, from FastAPI's side, I think the end solution to handle schema aliasing might be better placed in Pydantic; e.g. through an additional property in [model config](https://pydantic-docs.helpmanual.io/usage/model_config/).

**Comment by tiangolo** [comment, marked as answer]
Thanks for the discussion everyone.

`response_model_by_alias` is mainly a hack/workaround for quick experiments, mainly because some people were asking for that, but I'm inclined to deprecating and removing it at some point. Its behavior is actually quite strange, and there's a much better way to solve the same problem.

More background on that here: https://github.com/tiangolo/fastapi/pull/1642 and here: https://github.com/tiangolo/fastapi/pull/264

For your specific use case, your code example doesn't really run, I modified it to what I imagine is your intention:

```Python
from fastapi import FastAPI
from pydantic import BaseModel, Field

app = FastAPI()


class Transaction(BaseModel):
    id: str = Field(..., alias="_id")
    loan_id: str = Field(..., alias="loanId")


@app.get("/", response_model=Transaction, response_model_by_alias=False)
def read_root():
    return {"_id": "foo", "loanId": "bar"}
```

And when you call that endpoint, you would receive:

```json
{
  "id": "foo",
  "loan_id": "bar"
}
```

and then, you would want to have the JSON Schema inside OpenAPI declare the model with `id` and `loan_id`.

Now, notice that you return this in the function:

```python
{"_id": "foo", "loanId": "bar"}
```

but you received this in the HTTP JSON response:

```json
{
  "id": "foo",
  "loan_id": "bar"
}
```

**Note**: if you wanted to be able to return data using the field names and not the aliases you would need to [use `allow_population_by_field_name`](https://pydantic-docs.helpmanual.io/usage/model_config/).

## How to do it

Assuming something like that is what you want to achieve while having proper OpenAPI and JSON Schema support, this is how you would implement it:

```python
from fastapi import FastAPI
from pydantic import BaseModel, Field

app = FastAPI()


class Transaction(BaseModel):
    id: str = Field(..., alias="_id")
    loan_id: str = Field(..., alias="loanId")
    name: str  # to show that you can reuse fields with inheritance


class TransactionResponse(Transaction):
    id: str
    loan_id: str


@app.get("/", response_model=TransactionResponse)
def read_root():
    return {"id": "foo", "loan_id": "bar", "name": "baz"}
```

With this, you would receive in the JSON Response:

```json
{
  "id": "foo",
  "loan_id": "bar",
  "name": "baz"
}
```

And the JSON Schema would have `id` and `loan_id`:

![Selection_428](https://user-images.githubusercontent.com/1326112/117636464-208de600-b181-11eb-9e2b-e1241c04d709.png)

**Reply by YuriiMotov** [reply]
With Pydantic V2 just use `validation_alias` and `serialization_alias`:

```python
from fastapi import FastAPI
from pydantic import BaseModel, Field

app = FastAPI()

class TransactionResponse(BaseModel):
    id: str = Field(..., validation_alias="_id")
    loan_id: str = Field(..., validation_alias="loanId")
    name: str

@app.get("/", response_model=TransactionResponse)
def read_root():
    return {"_id": "foo", "loanId": "bar", "name": "baz"}
```

**Reply by foarsitter** [reply]
> With Pydantic V2 just use `validation_alias` and `serialization_alias`:
> 
> ```python
> from fastapi import FastAPI
> from pydantic import BaseModel, Field
> 
> app = FastAPI()
> 
> class TransactionResponse(BaseModel):
>     id: str = Field(..., validation_alias="_id")
>     loan_id: str = Field(..., validation_alias="loanId")
>     name: str
> 
> @app.get("/", response_model=TransactionResponse)
> def read_root():
>     return {"_id": "foo", "loanId": "bar", "name": "baz"}
> ```

This is the way to go!

**Comment by danmichaelo** [comment]
Thanks for providing the "How to do it" example. Can confirm that it works, but it does make the code more convoluted. It becomes even more convoluted when bringing SQL Alchemy and orm_mode into the mix. Building on the same example, I could not get something like this to work:

```python
@app.get("/", response_model=TransactionResponse)
def read_root():
    return db.query(TransactionModel).first()
```

One way I got it to work was like this:

```python
@app.get("/", response_model=TransactionResponse)
def read_root():
    transaction = db.query(TransactionModel).first()
    return TransactionResponse(**Transaction.from_orm(transaction).dict())
```

Please let me know if there is a less messy way!

Here's a complete self-contained example:

```python
from sqlalchemy import Column, Integer, String
from sqlalchemy.ext.declarative import declarative_base
from pydantic import BaseModel, Field
from fastapi import FastAPI, Depends


Base = declarative_base()


class TransactionModel(Base):
    __tablename__ = "transactions"
    _id = Column(Integer, primary_key=True)
    loanId = Column(String)
    name = Column(String)



class Transaction(BaseModel):
    id: str = Field(..., alias="_id")
    loan_id: str = Field(..., alias="loanId")
    name: str  # to show that you can reuse fields with inheritance

    class Config:
        orm_mode = True
        allow_population_by_field_name = True


class TransactionResponse(Transaction):
    id: str
    loan_id: str


app = FastAPI()

@app.get("/", response_model=TransactionResponse)
def read_root():
    transaction = TransactionModel(_id=1, loanId='ABC', name='Hello')

    return transaction  # <-- this does not work
    # return TransactionResponse(**Transaction.from_orm(transaction).dict())  # <-- this works, but messy
```

(Btw. In *my* case, the issue is that I need to return a field called "metadata", but "metadata" is reserved on SQL Alchemy models, so I use the workaround described [here](https://pydantic-docs.helpmanual.io/usage/models/#reserved-names))

**Reply by Ledorub** [reply]
Subclassing `GetterDict` may help. Response model in OpenAPI is rendered correctly.

```python
from pydantic import BaseModel
from pydantic.utils import GetterDict


class CustomGetter(GetterDict):
    def get(self, key: Any, default: Any = None) -> Any:
        if key == 'some_field_name':
            return getattr(self._obj, 'some_attribute', default)
        return super().get(key, default)


class Notification(BaseModel):
    some_field_name: str
    
    class Config:
        getter_dict = CustomGetter
```

I use this approach to flatten ORM objects.

**Reply by MahmudulHassan5809** [reply]
i am also facing the same issue when I fetch data from SQL Alchemy. 

```
from sqlalchemy import Column, String

from src.db.connector import Base


class User(Base):
    __tablename__ = "CRM_CUSTOMER"
    vcustid = Column(String, primary_key=True)
    vemail = Column(String, unique=True)

    # def __repr__(self):
    #     pass

```
This my Sql Alchemy model. and i can not change the field name cause this is not a new database. we are working on existing database


```
class UserBase(BaseModel):
    email: str = Field(..., alias="vemail")
    customer_id: str = Field(..., alias="vcustid")

class UserResponse(UserBase):
    email: str
    customer_id: str
    
    class Config:
        allow_population_by_field_name = True
        orm_mode = True
```
```
@router.get("/gold-balance", response_model=UserResponse)
@inject
def gold_balance(
    user_service: UserService = Depends(Provide[Container.user_service]),
):
    data = user_service.get_by_id()
    return data
```
got  fastapi.exceptions.ResponseValidationError 

@tiangolo any suggestion on how can I make it work? Thanks

**Reply by mhassan5809** [reply]
It is working now  
```
class UserBase(BaseModel):
    vemail: str = Field(..., alias="email")
    vcustid: str = Field(..., alias="customer_id")


class User(UserBase):
    class Config:
        populate_by_name = True
        from_attributes = True
```
after adding populate_by_name .

**Comment by jpribyl** [comment]
wish I had found this thread sooner! Looks like there is a pretty good workaround above but just in case it's helpful to anyone, I was able to patch `pydantic` so that it shows that it will accept the alias or the field

https://github.com/samuelcolvin/pydantic/pull/3005

**Comment by me-v2** [comment]
huge danger

**Comment by Bobronium** [comment]
Here's a hack that changes default behaviour to generate docs by field names instead of aliases:

```py
from functools import wraps
from unittest.mock import patch

from fastapi.utils import get_model_definitions
from pydantic.schema import model_process_schema


@wraps(model_process_schema)
def model_process_schema_wrapper(*args, **kwargs):
    kwargs.setdefault("by_alias", False)
    return model_process_schema(*args, **kwargs)


# Forces OpenAPI docs to have field names instead of aliases
# https://github.com/tiangolo/fastapi/issues/771
fastapi_schema_patch = patch(
    f"{get_model_definitions.__module__}.{model_process_schema.__name__}",
    model_process_schema_wrapper,
)
fastapi_schema_patch.start()
```

**Reply by rosscdh-tpg** [reply]
Unfortunately no longer valid in v2, not sure why this is so fiddley to get going, very simple use-case. Want input to match model fields not alias..

**Comment by Raphencoder** [comment]
Is there another way around besides @Bobronium solution ?

**Comment by rbuffat** [comment]
Pydantic allows choosing if the alias or the name should be used in the GenerateJsonSchema class.  There are not many code changes needed to expose this: https://github.com/tiangolo/fastapi/compare/master...rbuffat:fastapi:openapi_schema_without_alias?expand=1

**Comment by dsafanyuk** [comment]
using this also fixed it for me:

```python
# SQLAlchemy model
class Request(Base):
    __tablename__ = "Requests"

    ID = Column(BIGINT, primary_key=True)
    Status = Column(String)
    CreatedAt = Column(TIMESTAMP)
    LastModifiedAt = Column(TIMESTAMP, onupdate=func.utc_timestamp())


# Pydantic model
class Request(CreateRequest):

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: str = Field(validation_alias='ID', serialization_alias='id')
    status: StatusEnum = Field(validation_alias='Status', serialization_alias='status')
    created_at: str = Field(validation_alias='CreatedAt', serialization_alias='created_at')
    last_modified_at: str = Field(validation_alias='LastModifiedAt', serialization_alias='last_modified_at')
```

**Comment by mmzeynalli** [comment]
@dsafanyuk that option is no go for me: I have a parent pydantic class which has `alias`ed fields, it is the child model that decides to use aliased name or not with ConfigDict:

```python
class MonthYearToDateSchema(BaseModel):
    start_date_month: int = Field(
        exclude=True,
        alias=AliasChoices('start_date_month', 'issue_date_month'),
    )
    start_date_year: int = Field(
        exclude=True,
        alias=AliasChoices('start_date_year', 'issue_date_year'),
    )
    
    @computed_field(alias='issue_date')  # type: ignore[misc]
    @property
    def start_date(self) -> date:
        return date(self.start_date_year, self.start_date_month, 1)
    
class UserWorkExperienceSchema(DateToMonthYearSchema):
    model_config = ConfigDict(populate_by_name=False)
    id: int
    currently_working: bool
    
class CertificateSchema(DateToMonthYearSchema):
    model_config = ConfigDict(populate_by_name=True)
    id: int
```

I have these two models, which are both field type of `UserProfile` model:

```python
class UserProfileSchema(BaseModel):
    work_experiences: list[UserWorkExperienceSchema] = []
    certificates: list[CertificateSchema] = []
```

Even though, I want `start_date` for `UserWorkExperience` and `issue_date` for `UserCertificate` to be serialized, I get either of those for both models, even though I have defined different values for `populate_by_name`, because of     `response_model_by_alias` attribute of APIRoute.

**Comment by wallneradam** [comment]
This is my easy monkey patching solution for this problem:

```python
import pydantic.json_schema

_orig_generate_json_schema_init = pydantic.json_schema.GenerateJsonSchema.__init__


def _new_generate_json_schema_init(self, by_alias=False, *args, **kwargs):
    _orig_generate_json_schema_init(self, by_alias, *args, **kwargs)


pydantic.json_schema.GenerateJsonSchema.__init__ = _new_generate_json_schema_init
```

This is similar to @rbuffat `s solution, but you don't need to change anything inside FastAPI. It just overrides the default of Pydantic.

**Reply by rosscdh-tpg** [reply]
Perfect for v2 thank you very much for this, tbh this is a fairly obvious bit of functionality when dealing with integration with older less consistent systems. 

Thanks for providing it

**Comment by nuno-andre** [comment]
IMO @rbuffat's proposal is on the right track. But maybe it would better to enable the option to change the whole generator class instead of passing such a specific field.

```python
class FastAPI(Starlette):
    def __init__(
        ...
        json_schema_generator: type(GenerateJsonSchema) = GenerateJsonSchema,
        ...
    ):
```

And then, as proposed by @wallneradam, simply pass a `GenerateJsonSchema` subclass:

```python
class ByFieldJsonSchema(GenerateJsonSchema):
    def __init__(self, *args, **kwargs):
        kawrgs['by_alias'] = False
        super().__init__(*args, **kwargs)
```

https://docs.pydantic.dev/latest/api/json_schema/#pydantic.json_schema.GenerateJsonSchema

¿Qué opinas, @tiangolo?

---

A kludgy alternative, which doesn't involve monkey-patching but does the job, is overriding the `__get_pydantic_json_schema__` model method (which is called by `GenerateJsonSchema` during docs generation):


```python
from pydantic import BaseModel

class ByFieldBaseModel(BaseModel):

    @classmethod
    def __get_pydantic_json_schema__(cls, core_schema, handler):

        def alias_to_field(schema):
            if 'schema' in schema:
                alias_to_field(schema['schema'])
            elif 'fields' in schema:
                for field in schema['fields']:
                    schema['fields'][field]['serialization_alias'] = field

        alias_to_field(core_schema)
        return handler(core_schema)
```
